### PR TITLE
Do not force display render for software cursor restorer: fix render of incorrect frames after closing game dialogs with software cursor enabled

### LINKED ISSUE #7272: Program, closing some screens may cause a different mouse cursor to appear briefly
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

In fheroes2, closing some screens may cause a different mouse cursor to appear briefly.

In fheroes2:

https://github.com/ihhub/fheroes2/assets/43583994/1f361a2d-b500-46dc-8888-f7d41d4a973f

Shotcut_00_00_08_280

### Save file

Here is the save file:
[Dominion_0064.zip](https://github.com/ihhub/fheroes2/files/11615881/Dominion_0064.zip)

### Additional info

Version 1.0.4 7265

Related issues: https://github.com/ihhub/fheroes2/issues/6937, https://github.com/ihhub/fheroes2/issues/1266.

### LINKED ISSUE #7281: Opening dialogs from battle options dialog causes short flicker
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

I believe there is some unnecessary redrawing happening when opening either the audio options or hot keys dialog from the battle options as there is a short moment where you can view the background changing before the new dialog is opened.

I haven't had time to investigate exactly what happens but it was only recently I noticed it. I believe it is even more noticeable on slower devices.

Play the video slowly to hopefully see it. It's somewhat more noticable when opening the audio settings. Also the video is reduced to 30 fps or lost some frames so I see that it isn't as easy to spot on it.

https://github.com/ihhub/fheroes2/assets/12501091/e567da77-acea-47d3-a507-6b97ad9f0641

### Save file

Just open the battle only mode.

### Additional info

_No response_

### LINKED ISSUE #7730: VSync ON causes flickering fade out when exiting certain screens back to the adventure map
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.
- [X] I've checked that this issue is reproducible on the [**latest snapshot build**](https://ihhub.github.io/fheroes2/INSTALL.html#snapshots-latest-builds).

### Platform

Windows

### Describe the bug

When exiting back to the adventure map I sometimes observe a fade out that flickers a lot, but only with Vsync ON.

It happens in all the cases below when going from hero screen, castle screen, hero meeting screen. But, it does not happen when you leave the kingdom overview screen or apparently the hero screen if you enter it by pressing the mini hero portrait on the side bar, same with the castle screen if you enter it from the side bar.

https://github.com/ihhub/fheroes2/assets/12501091/d6ded272-94db-41a7-a95d-d4bbf0f1d3fc

### Save file

No save file needed

### Additional info

Software cursor ON
VSync ON
Nearest filtering

### PR DESCRIPTION
**Pull request by Districh-ru**
Fix #7730, fix #7281, fix #7272

This PR removes the forced display `render()` call from `CursorRestorer` destructor which sometimes caused render of not finished incorrect frames when software cursor is turned on. And fixes delayed update of the cursor image which caused incorrect cursor after closing some dialogs.

The cursor change logic for the Game Area now is put right after left/right click events processing to immediately update the cursor with the proper image.

This PR also modifies the logic after the end of hero's movement including interaction with game objects:
In master build the hero moving state is updated (along with the cursor) only on the next loop (for the next screen frame) after the hero ends his move.
In this PR the hero's "ENABLEMOVE" state is reset right after the end of hero's move along with updating the cursor.

### CODE REVIEW

**Comment by Districh-ru** [comment]
Hello, @LeHerosInconnu. Could you please test this PR for issue #7272 if you have time.

**Comment by Districh-ru** [comment]
> Hi, @Districh-ru !
> 
> I've caught one issue during testing of this PR.

Hi, @Branikolog, many thanks for testing this PR!
Good catch! I played for 2 hours and did not notice this bug... :(
It is fixed now, could you please check it when you have time.

**Comment by LeHerosInconnu** [comment]
Hello @Districh-ru,

> Hello, @LeHerosInconnu. Could you please test this PR for issue #7272 if you have time.

So I tested this and the problem described in issue #7272 is still present.

https://github.com/ihhub/fheroes2/assets/43583994/a7a07738-91cb-4be2-b335-fd30b4f3a786

Shotcut_00_00_10_760

**Comment by Districh-ru** [comment]
Hi, @LeHerosInconnu, many thanks for testing and for the video with the issue!

I found another place which caused the cursor change in the middle of fade animation. 
Now in such dialogs with fade effect the cursor restorer is removed. The cursor should now stay unchanged until fade animation finishes and the Game Area events handler will set the appropriate cursor.

Could you please take a look when you have time.

Cursor restorer is not needed in these dialogs (and probably in most others) because after closing the dialog, opened over the game area or the battlefield, the events handler will replace the restored cursor with the appropriate one. The cursor restoration is redundant here and also a little harmful because it restores the previous cursor image that is not always correct for the new cursor position.

**Comment by LeHerosInconnu** [comment]
Hello @Districh-ru,

> Hi, @LeHerosInconnu, many thanks for testing and for the video with the issue!
> 
> I found another place which caused the cursor change in the middle of fade animation. Now in such dialogs with fade effect the cursor restorer is removed. The cursor should now stay unchanged until fade animation finishes and the Game Area events handler will set the appropriate cursor.
> 
> Could you please take a look when you have time.
> 
> Cursor restorer is not needed in these dialogs (and probably in most others) because after closing the dialog, opened over the game area or the battlefield, the events handler will replace the restored cursor with the appropriate one. The cursor restoration is redundant here and also a little harmful because it restores the previous cursor image that is not always correct for the new cursor position.

I tested this quickly with [Dominion_0064.zip](https://github.com/ihhub/fheroes2/files/11615881/Dominion_0064.zip) save file and the problem reported in issue #7272 seems to have been solved.
I'll keep an eye on this behavior in future games. :)
Thanks for the fix.

**Comment by Districh-ru** [comment]
Hello, @LeHerosInconnu. 
> I tested this quickly with [Dominion_0064.zip](https://github.com/ihhub/fheroes2/files/11615881/Dominion_0064.zip) save file and the problem reported in issue #7272 seems to have been solved.
I'll keep an eye on this behavior in future games. :)
Thanks for the fix.

I am glad to hear it. :)
Thank you for testing!

**Review by Branikolog — APPROVED** [review]

File: src/fheroes2/battle/battle_dialogs.cpp [hunk 1]
```diff
@@ -248,8 +248,9 @@ namespace
         LocalEvent & le = LocalEvent::Get();
         Settings & conf = Settings::Get();
 
-        // setup cursor
-        const CursorRestorer cursorRestorer( true, Cursor::POINTER );
+        // Set the cursor image.This dialog is called from the battlefield and does not require a cursor restorer.
```
Comment (ihhub) [inline src/fheroes2/battle/battle_dialogs.cpp:251]: Please add a white space after the first full stop.

File: src/fheroes2/battle/battle_dialogs.cpp [hunk 2]
```diff
@@ -248,8 +248,9 @@ namespace
...
+        // Battlefield event proccessror will set the appropriate cursor after this dialog is closed.
```
Comment (ihhub) [inline src/fheroes2/battle/battle_dialogs.cpp:252]: **proccessror** is misspelled.

File: src/fheroes2/battle/battle_dialogs.cpp [hunk 3]
```diff
@@ -729,8 +730,8 @@ bool Battle::Arena::DialogBattleSummary( const Result & res, const std::vector<A
 
 void Battle::Arena::DialogBattleNecromancy( const uint32_t raiseCount )
 {
-    // setup cursor
-    const CursorRestorer cursorRestorer( true, Cursor::POINTER );
+    // Set the cursor image.This dialog does not require a cursor restorer.
```
Comment (ihhub) [inline src/fheroes2/battle/battle_dialogs.cpp:733]: Please add a white space after the first full stop.

File: src/fheroes2/castle/castle_dialog.cpp [hunk 4]
```diff
@@ -217,8 +217,9 @@ namespace
 
 Castle::CastleDialogReturnValue Castle::OpenDialog( const bool openConstructionWindow, const bool fade, const bool renderBackgroundDialog )
 {
-    // setup cursor
-    const CursorRestorer cursorRestorer( true, Cursor::POINTER );
+    // Set the cursor image.This dialog does not require a cursor restorer. It is called from other dialogs that have the same cursor
```
Comment (ihhub) [inline src/fheroes2/castle/castle_dialog.cpp:220]: Please add a white space after the first full stop.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @Districh-ru , I left very minor comment related remarks. Could you please take a look at them in your free time?

**Review by ihhub — APPROVED** [review]

File: src/fheroes2/battle/battle_dialogs.cpp (hunk 1 above)
Comment (Districh-ru) [reply src/fheroes2/battle/battle_dialogs.cpp]: Hi, @ihhub. Thanks for the spell checking!

File: src/fheroes2/game/game_startgame.cpp [hunk 5]
```diff
@@ -1255,14 +1255,41 @@ fheroes2::GameMode Interface::AdventureMap::HumanTurn( const bool isload )
...
                         }
 
-                        isMovingHero = true;
+                        // Update the hero's move status.
+                        isMovingHero = hero->isMoveEnabled();
 
                         if ( hero->isAction() ) {
-                            // check if the game is over after the hero's action
+                            // The action can not be performed while moving, only after the move is ended.
+                            assert( !isMovingHero );
```
Comment (ihhub) [inline src/fheroes2/game/game_startgame.cpp:1263]: My only worry with this assert is that when a moving hero steps on an event. Could you please verify that it doesn't assert for this case?

Comment (Districh-ru) [reply src/fheroes2/game/game_startgame.cpp:1263]: I checked, its OK.
When event is occurred the hero stops. And the event action is performed in `MoveStep()` called from `hero->Move()`. So at this moment the assertion will not blow up and the condition `hero->isAction()` will be false.

**Comment by ihhub** [comment]
@Districh-ru , huge thanks for this awesome work!

**Pull request merged**